shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1102,29 +1114,105 @@ - (void)keyboardWillChangeFrame:(NSNotification*)notification {
+  UIViewAnimationCurve curve =
+      (UIViewAnimationCurve)[[info objectForKey:UIKeyboardAnimationCurveUserInfoKey] integerValue];
+
   // Considering the iPad's split keyboard, Flutter needs to check if the keyboard frame is present
   // in the screen to see if the keyboard is visible.
   if (CGRectIntersectsRect(keyboardFrame, screenRect)) {
     CGFloat bottom = CGRectGetHeight(keyboardFrame);
     CGFloat scale = [UIScreen mainScreen].scale;
-
     // The keyboard is treated as an inset since we want to effectively reduce the window size by
     // the keyboard height. The Dart side will compute a value accounting for the keyboard-consuming
     // bottom padding.
-    _viewportMetrics.physical_view_inset_bottom = bottom * scale;
+    self.targetViewInsetBottom = bottom * scale;
   } else {
-    _viewportMetrics.physical_view_inset_bottom = 0;
+    self.targetViewInsetBottom = 0;
   }
-
-  [self updateViewportMetrics];
+  [self startKeyBoardAnimation:duration curve:curve];
 }
 
 - (void)keyboardWillBeHidden:(NSNotification*)notification {
-  _viewportMetrics.physical_view_inset_bottom = 0;
-  [self updateViewportMetrics];
+  // When keyboard hide,the keyboardWillChangeFrame function will be called to update viewport
+  // metrics. So do not call [self updateViewportMetrics] here again
+}
+
+- (void)startKeyBoardAnimation:(NSTimeInterval)duration curve:(UIViewAnimationCurve)curve {
+  // If current physical_view_inset_bottom == targetViewInsetBottom,do nothing
+  if (_viewportMetrics.physical_view_inset_bottom == self.targetViewInsetBottom) {
+    return;
+  }
+
+  // When call this method first time,
+  // initialize the keyboardAnimationView to get animation interpolation during animation
+  if ([self keyboardAnimationView] == nil) {
+    _keyboardAnimationView.reset([[UIView alloc] init]);
+    [self.view addSubview:[self keyboardAnimationView]];
+  }
+
+  if (self.keyboardAnimationLink == nil) {
+    [self startKeyboardAnimationLink];
+  }
+
+  // Remove running animation when start another animation,and after call this line,the `completion`
+  // in UIView animation will be called. If it is animating,the `finished` will be false,and the
+  // `keyboardAnimationLink` will not invalidate.When animation finishes.The `finished` will be true
+  // and `keyboardAnimationLink` will invalidate
+  [[self keyboardAnimationView].layer removeAllAnimations];
+
+  // Set animation begin value
+  [self keyboardAnimationView].frame =
+      CGRectMake(0, _viewportMetrics.physical_view_inset_bottom, 0, 0);
+
+  [UIView animateWithDuration:duration
+      animations:^{
+        // Set end value
+        [self keyboardAnimationView].frame = CGRectMake(0, self.targetViewInsetBottom, 0, 0);
+      }
+      completion:^(BOOL finished) {
+        // Only when animation finished ,we invalidate the `keyboardAnimationLink` and
+        //  set end value to _viewportMetrics
+        if (finished) {
+          [self stopKeyboardAnimation];
+        }
```

Comment:
This is an asynchronous call, so you could call `removeAllAnimations` and this callback could be called with `YES`.  There is no guarantee I know of that says `completion` will be called synchronously with `removeAllAnimations`.  In that case we'd be waiting for a CADisplayLink to run that doesn't exist.  I think it's safe just to not try to reuse a CADisplayLink.  Just kill the old one and start a new one.

We might as well remove the UIView when we cleanup too.

Reply:
done